tests/integration_tests/instances.py

```diff
@@ -43,21 +52,37 @@ def emit_settings_to_log(self) -> None:
     def destroy(self):
         self.instance.delete()
 
-    def execute(self, command):
-        return self.instance.execute(command)
+    def execute(self, command, *, use_sudo=None) -> Result:
```

Comment:
One potential gap here: test authors may use `use_sudo=False` to mean "run this unprivileged". If we have a case where we run commands on instances as root by default, then it will not have that effect: it will run as root regardless of the value of the parameter.

Given that this will only happen if `instance.username` is "root", and we have no immediate expectation that it will take that value, I wonder if a check that raises an `Exception("You need to implement this functionality now, sorry about that.")` if `self.instance.username == "root" and not use_sudo` might be the most effective way to (a) not solve that problem until we need to, and (b) get test failures that prompt us to solve that problem once we do need to.

What do you think?

Reply:
Heh...that'd be a funny exception to get :smile: 

If https://github.com/canonical/pycloudlib/pull/56 lands, then I think all pycloudlib instances will run unprivileged by default, so this becomes a non-issue. It seems to be low enough risk that I don't feel compelled to change it at this point.

Reply:
> all pycloudlib instances will run unprivileged by default, so this becomes a non-issue

All pycloudlib instances will _not use sudo_ by default, but if the default SSH user is `root` (which it may well be in some clouds/images) then that is not the same as running unprivileged. The subtlety of the mismatch is exactly why I want an explicit warning raised, because I doubt that we'll notice it when it actually comes up.

(And because this is such a corner case, that's why I'm proposing we just add two LOC to raise an exception indicating that someone has hit it; it's not worth investing time until we actually need to resolve it.)

Reply:
I added the check, but as currently implemented LXD won't hit this. The instance username is set to `ubuntu` everywhere, and that determines the user to ssh in as.

Reply:
Yep, we should only hit this at some point in the future, possibly never. Thanks!